lib/ref_arch_setup/install.rb

```diff
@@ -25,61 +31,307 @@ def initialize(target_master)
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The PE tarball URL
+    #
+    # @return [string] The tarball path on the target master
+    def handle_tarball_url(url)
+      parse_url(url)
+      tarball_path_on_master = "#{TMP_WORK_DIR}/#{@pe_tarball_filename}"
+      remote_error = "Failed downloading #{url} to localhost and moving to #{@target_master}"
+
+      if target_master_is_localhost?
+        success = download_pe_tarball(url, "localhost")
+        raise "Failed downloading #{url} to localhost" unless success
+      else
+        # if downloading to the target master fails try to download locally and then upload
+        success = download_pe_tarball(url, @target_master)
+        puts "Unable to download the tarball directly to localhost" unless success
+        success = download_and_move_pe_tarball(url) unless success
+        raise remote_error unless success
+      end
+
+      return tarball_path_on_master
+    end
+
+    # Copies the PE tarball from the specified location to the temp working directory
+    # on the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The pe tarball path
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def copy_pe_tarball_on_target_master(path)
+      command = "cp #{path} #{TMP_WORK_DIR}"
+      success = BoltHelper.run_cmd_with_bolt(command, @target_master)
+      return success
+    end
+
+    # Handles the specified tarball path when the target master is not localhost
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The pe tarball path
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def handle_tarball_path_with_remote_target_master(path)
+      remote_flag = "#{@target_master}:"
+
+      if path.start_with?(remote_flag)
+        actual_path = path.sub!(remote_flag, "")
+        success = file_exist_on_target_master?(actual_path)
+        success = copy_pe_tarball_on_target_master(actual_path) if success
+      else
+        success = File.exist?(path)
+        success = upload_pe_tarball(path) if success
+      end
+
+      return success
+    end
+
+    # Handles the specified tarball path
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The PE tarball path
+    #
+    # TODO: improve "host to install on"? ("host where PE will be installed?")
+    #
+    # @return [string] The tarball path on the target master
+    def handle_tarball_path(path)
+      filename = File.basename(path)
+      tarball_path_on_master = "#{TMP_WORK_DIR}/#{filename}"
+      file_not_found_error = "File not found: #{path}"
+      upload_error = "Unable to upload tarball to the RAS working directory on #{@target_master}"
+      copy_error = "Unable to copy tarball to the RAS working directory on #{@target_master}"
+
+      if target_master_is_localhost?
+        raise file_not_found_error unless File.exist?(path)
+        error = copy_error
```

Comment:
This reads badly.  Seeing this line feels like an error has happened when it hasn't.  There are a couple of ways to change that... Could just do error = copy_error unless upload_pe...   and then later raiser error unless error.nil?    Also could just change the variable name to potential_error.

Reply:
Updated to:
```
      if target_master_is_localhost?
        raise file_not_found_error unless File.exist?(path)
        success = upload_pe_tarball(path)
        error = copy_error unless success
      else
        success = handle_tarball_path_with_remote_target_master(path)
        error = upload_error unless success
      end

      raise error unless success
```